src/pybind/mgr/volumes/fs/operations/volume.py

```diff
@@ -117,6 +117,58 @@ def delete_volume(mgr, volname, metadata_pool, data_pools):
     result_str = "metadata pool: {0} data pool: {1} removed".format(metadata_pool, str(data_pools))
     return r, result_str, ""
 
+def rename_volume(mgr, volname, newvolname, rename_pools):
+    """
+    rename volume (file system and its pools)
+    """
+    metadata_pool, data_pools = get_pool_names(mgr, volname)
+    new_metadata_pool, new_data_pool = gen_pool_names(newvolname)
+
+    if rename_pools:
+        if len(data_pools) > 1:
+            return (
+                -errno.EINVAL,
+                "",
+                f"Failed to rename pools of volume '{volname}'. It has more "
+                "than one data pool. Rename its pools manually.")
+        data_pool = data_pools[0]
+
+        r, outb, outs = rename_pool(mgr, metadata_pool, new_metadata_pool)
+        if r != 0:
+            log.error(
+                f"Failed to rename metadata pool '{metadata_pool}' of volume "
+                "'{volname}' to '{new_metadata_pool}'")
+            return r, outb, outs
+
+        r, outb, outs = rename_pool(mgr, data_pool, new_data_pool)
+        if r != 0:
+            log.error(
+                f"Failed to rename data pool '{data_pool}' of volume "
+                "'{volname}' to '{new_data_pool}'")
+            r, outb, outs = rename_pool(mgr, new_metadata_pool, metadata_pool)
+            if r != 0:
+                log.error(
+                    f"Failed to rename metadata pool '{new_metadata_pool}' of "
+                    "volume '{volname}' to original name '{metadata_pool}'")
+            return r, outb, outs
+
+    r, outb, outs = rename_filesystem(mgr, volname, newvolname)
```

Comment:
I think this needs to be a stack based approach: file system creation is last when doing `volume create` so file system rename should be first when doing `volume rename`.

Reply:
Done